service_capacity_modeling/capacity_planner.py

```diff
@@ -576,6 +576,58 @@ def _plan_certain(  # pylint: disable=too-many-positional-arguments
         num_results = num_results or self._default_num_results
         return reduce_by_family(plans)[:num_results]
 
+    # Calculates the minimum cpu, memory, and network requirements based on desires.
+    def _per_instance_requirements(self, desires) -> Tuple[int, float, float]:
+
+        # Applications often set fixed reservations of heap or OS memory
+        per_instance_mem = (
+            desires.data_shape.reserved_instance_app_mem_gib
+            + desires.data_shape.reserved_instance_system_mem_gib
+        )
+
+        # Applications often require a minimum amount of true parallelism
+        per_instance_cores = int(
+            math.ceil(
+                desires.query_pattern.estimated_read_parallelism.mid
+                + desires.query_pattern.estimated_write_parallelism.mid
+            )
+        )
+
+        per_instance_net = 0.0
+
+        current_capacity = (
+            None
+            if desires.current_clusters is None
+            else (
+                desires.current_clusters.zonal[0]
+                if len(desires.current_clusters.zonal)
+                else desires.current_clusters.regional[0]
+            )
+        )
+        # Return early if we dont have current_capacity set.
+        if current_capacity is None or current_capacity.cluster_instance is None:
+            return (per_instance_cores, per_instance_mem, per_instance_net)
+
+        # Calculate CPU requirements based on current capacity
+        current_cpu_utilization = current_capacity.cpu_utilization.high / 100.0
+        current_instance_cpus = current_capacity.cluster_instance.cpu
+        required_cores = math.ceil(current_cpu_utilization * current_instance_cpus)
```

Comment:
This seems odd to me, since the model can return a higher number of smaller computer they could get to this utilization number a different way (e.g. 2x computers would halve the utilization)

Reply:
Hm I stilll think this is a tad odd to filter on this, since we could achieve the utilization with fewer cores (but more machines) - but I do agree it's conservative as when rightsizing we would never go up in per machine utilized cores

So if you have 100 4xl with 50% utilization, thats 16 * 8 minimum 8 vcpus, imo it would be fine to have 200 2xl (which only have 8 cores total per box

Reply:
@jolynch I think this calculation adheres to your example. The filter would indeed resolve to min 8vcpus which would indeed allow 2xls through. It would filter out instances with less than 8vcpus.

Reply:
One potential issue is that we may need to normalize these, but Ill leave that for a follow up.

Reply:
I think a more extreme example. If we have a current cluster of 100 * m5.8xl @ 80%, and scale linearly, it could theoretically run 800 * m5.xl @ 80%. If I read correctly, I think this would come up with `current_util 80% * current_cpus = 32 = 25cores`, which would exclude anything below 8xl IIUC?

Reply:
Hmm yes that does help. I think my mental model is incorrect, I have indeed been treating cpus as "incompressible" as Joey likes to put it.

It does seem wrong to think we could go to the extreme and run on a bunch of 1vcpu instances, but my calculation here is likely much too strict. Maybe I should remove it here and just leave it up to the model then.

Reply:
Removed filtering on cpu utilization.